### ISSUE TITLE
Should we keep Stardog support?

### ISSUE THREAD

**Issue by kal** [dotnetrdf/dotnetrdf#346]
Stardog is no longer available in the free Community edition and so to set up a test environment for it will require a license (the free license now only lasts for 30 days and has to be applied for). This will make ongoing testing / support for Stardog difficult. 

I want to check with contributors to the Stardog connector what they want to do about this situation. I think that the options are:

1) Remove the Stardog connectors (maybe mark them as obsolete in the 3.0 release and remove them later)
2) Find a volunteer with a Stardog license to take over maintenance of the connector
3) as (2) but also move the Stardog connector into its own library so that the maintainer can have a bit more control over the release cycle

@charbull  @DarthStem - tagging you two as you have provided contributions to those connectors and if either of you have the infrastructure and the willingness to take on the maintainance that would be great!

@tpluscode @rvesse @langsamu - tagging you as you also have a stake in all this and I'd appreciate your feedback.

Also genuinely interested in other ideas for how to deal with this situation.

**Comment by langsamu** [comment]
SL1. Disclaimer: I have no technical or community experience with Stardog, nor am I familiar with the relevant components in dotnetrdf.
SL2. I think a major release 3.0 is a good point to remove a feature completely if so decided (rather than marking as obsolete).
SL3. My experience with other vendors leads me to believe they might be open to granting us limited license sufficient for development/testing. But again, I have no idea about Stardog specifically.
SL4. As a contributor I support moving vendor specific functionality to its own library as per KA2. In addition to decoupling release cycles this would also reduce our own codebase and help improve our API.
SL5. I don't think we should keep any code that relies on specific contributors having development/testing infrastructure.

**Comment by joephayes** [comment]
I work for Stardog and I'd like to participate in this discussion. Can I get some additional details on what the separate library model look like? Are there examples of externally maintained connectors you can point me to?

**Comment by kal** [comment]
Thanks @langsamu - there are some good points there. 

> SL2. I think a major release 3.0 is a good point to remove a feature completely if so decided (rather than marking as obsolete).

I had started down that route and then decided to try and see if I could stick to the old dotNetRDF deprecation policy of deprecating in one release and removing in the next to help minimize the transition burden. That said, there is quite a lot of old API that needs to be updated (especially around the handling of async operations which has been vastly improved in .NET/C# since the original DNR async code was written).

> As a contributor I support moving vendor specific functionality to its own library as per KA2. In addition to decoupling release cycles this would also reduce our own codebase and help improve our API by excercising what's exposed and how.

This makes sense, though with the additional thought that some of those connectors could be maintained within the scope of this project and just packaged as separate assemblies/nuget packages. Again this would be a bit of a breaking change (though easily fixed if it is just a question of referencing an additional nuget package).

> I don't think we should keep any code that relies on specific contributors posessing unique development/testing infrastructure.

I'm currently working (locally, not committed yet) on writing Dockerfiles for the different backends to make it possible for any contributor to set up a test environment with minimum fuss - that is in fact where this issue regarding Stardog came up as I realized that it would be hard to automate a build or provide a test image.

**Comment by kal** [comment]
Thanks for reaching out @joephayes !

Currently the vast majority of the connectors are part of the dotNetRDF codebase. The only externally maintained one that I'm aware of is the connector for Blazegraph (https://github.com/blazegraph/blazegraph-dotnetrdf).

There are a two sets of interfaces for connectors to implement. The `VDS.RDF.Storage.IStorageProvider` (and its companions) that provide access to a single store (KB) and support operations such as import, export, query, update etc. There's a collection of interfaces that allow you to incrementally add feature support. The `VDS.RDF.Storage.Management.IStorageServer` provides an admin API for listing and managing stores and there is a companion `VDS.RDF.Storage.Management.Providioning.IStoreTemplate` interface that can be used for templated configuration of new stores.

https://github.com/dotnetrdf/dotnetrdf/wiki/UserGuide-Storage-Providers is a good jumping off point for seeing what we have in the documentation around this area.

Please note that the Async versions of those interfaces are all callback-based as they predate Task and async/await. My plan is to deprecate those methods and add a more modern async interface as part of the work on release 3.0.

**Comment by joephayes** [comment]
> I'm currently working (locally, not committed yet) on writing Dockerfiles for the different backends to make it possible for any contributor to set up a test environment with minimum fuss - that is in fact where this issue regarding Stardog came up as I realized that it would be hard to automate a build or provide a test image.

A solution can be to provide a Dockerfile / docker-compose file that mounts the license from an external directory into the container. You still need to acquire the license, but it enables you to ship a generic Dockerfile...

**Comment by kal** [comment]
That's a possibility - and it would definitely fix the more immediate issue for me working on the code right now. However it is an extra bit of friction for any future maintainer and as I understand it the evaluation period is time limited so although I could get a license to use now, it would potentially be a problem in a couple of months should I need to go back to the code. Also I would like to use the docker images in CI builds which could be problematic with license files.

Sorry to sound negative! Please don't get me wrong, I understand that you guys need to protect your software! I've used Stardog in a couple of projects and have enjoyed working with it and if we can find a good way to keep a Stardog connector in dotNetRDF I'd definitely like to do that.

**Comment by joephayes** [comment]
@kal - just wanted to mention that Stardog now offers a free version - See https://www.stardog.com/pricing/ and https://www.stardog.com/get-started/

**Comment by kal** [comment]
Thanks @joephayes ! I've got myself a license file and done a bit of work on updating the tests for the 3.0 release. The tests won't be part of our CI build as I don't want to push a license file to Github and I'm not sure that there is a way to have a secret file in Appveyor. I'm happy with that as a compromise solution though as it is still pretty easy for me to run tests locally and now I don't have a finite evaluation period to worry about 👍 

Going to close this ticket now, but if you guys want to talk some more about deepening or broadening the support for Stardog in dotNetRDF, please don't hesitate to get in touch.

**Comment by kal** [comment]
TL;DR - yes we will keep Stardog support :)
